### DISCUSSION TITLE
Fork: UnamWebPanelPlus release v0.1.0

### DISCUSSION THREAD

**Discussion by alcinzal** [UnamSanctam/UnamWebPanel#327] [Show and tell]
<img src="https://github.com/Alcinzal/UnamWebPanelPlus/blob/master/UnamWebPanelPlus.png?raw=true">

# UnamWebPanelPlus v0.4.0

I have released a modified version of the UnamWebPanel with a few extra features. You can see it here: [UnamWebPanelPlus](https://github.com/Alcinzal/UnamWebPanelPlus/)

For UnamSanctam: If you see anything you like, feel free to implement it into your own project. No credit is needed 😊. No pressure either of course if you feel the features are impractical or not worth implementing.

Any feedback is appreciated!
## Differences
* Lifetime statistics - Statistics are saved every hour forever. You can view the statistics and toggle between viewing the data hourly, daily, weekly or monthly.
* Geo - See which country and continent your miners are from in a pie chart. Country is also visible in the datatable.
* Total Hashrate - An overview of each algorithm, and their total hashrate, online miners and their average hashrate.
* Cloudflare support - If you use Cloudflare as a DNS server, the panel will still get the correct IP address and country code, from the headers 'CF-Connecting-IP' and 'CF-IPCountry'.
* Better timezone support - All dates are saved as Unix Timestamps in the database, ensuring that dates stay true no matter what timezone. (In the future you will be able to manually select timezone on a settings page).
* More to come!

### PLUS 0.4.0 (2024-02-10)
* Merged this fork with Unams 1.8.0 update.
    * Changed almost all dates to Unix Timestamp (including database dates).
    * Changed all charts to display date in 24-hour format.
    * Integrated cloudflare support.
    * Integrated plus page.
    * Integrated total hashrate functionality.
    * Integrated statistics functionality.
        * Integrated statistics table to database.
        * Removed tooltips in the statistics viewer.
    * Integrated geolocation functionality.
        * Integrated country column to database
    * Did not integrate my previous configuration modifications as I plan to redesign this entire page in the future.
    * Did not integrate my previous XSS prevention attempts as 1.8.0 renders it obsolete.
* Added some branding/identity to my plus fork (for fun).
    * Added a more detailed drawing of Unam.
    * Changed "UnamWebPanel" to "UnamWebPanel+".
* Changed terminology: "Active" -> "Mining (Active)", "Idle" -> "Mining (Idle)", Both of these combined -> "Mining".
### PLUS 0.3.7 (2024-02-05)
* Fixed XSS string checking.
### PLUS 0.3.6 (2024-02-02)
* Replaced character limit with blacklisted string checking, to better fight off XSS attacks.
### PLUS 0.3.5 (2024-01-31)
* Added character limit to algorithms so XSS attempts do not go through.
* Fixed small bug where statistics would throw an error if there was no statistics data.
* Changed total hashrate cards so yellow means active miners but no hashrate, and green means active miners with hashrate.
### PLUS 0.3.0 (2024-01-23)
* Changed dates in database to Unix Timestamp for better compatibility with changing timezones.
* Added Cloudflare support. Checks for headers 'CF-Connecting-IP' and 'CF-IPCountry'.
* Changed formatting in unamtChart to use a 24 hour clock instead of 12 hour.
* Changed countries to Geo.
* Added continents to Geo.
* Saved country names and continent names as json files, to prevent calling a static json from another server (country.io).
* Optimized the Plus statsSaver file.
* Changed the definition of online and offline miners. Offline meaning no connection for 180 seconds, and online meaning the opposite, no matter the status.
* Optimized the displaying of statistics at the Plus page.
* Fixed bug where 2023-12-31 was week 1. Now it is correct, showing week 53.
### PLUS 0.2.5 (2024-01-16)
* Fixed bug where statistics graphs would disappear.
* Fixed the charts not updating when the resizing the browser.
* Optimized code a little to not call the same files multiple times.
* Fixed non-existent file being called.

[You can view the full changelog of this fork here](CHANGELOG.md)

**Comment by UnamSanctam** [comment]
Great job! For the next version I am rewriting a lot of the code to make it easier for others to edit.

**Reply by alcinzal** [reply]
Thank you! That sounds great, I will try to see what I can contribute with when that time comes 😁

**Reply by alcinzal** [reply]
Hi again. I made an update and actually managed to get everything working nicely with the database, instead of using json files, but I just have 2 questions for you.

#### 1. bootstrap.custom.css error
I keep getting the error that "assets/css/bootstrap.custom.css" was not found, and I don't really see it in my own or your repository. It gets called in templates.php at "generalCSSIncludes()", can I just remove this? Or is it important?

#### 2. Configurations
I am trying to make the configurations page easier to use and understand, so I was wondering about the settings. From the wiki this is what I was able to gather about what settings fit where. I have splitted them up a little for easier viewing.

##### CPU Miner
```
{
    "algo": "",
    "pool": "",
    "port": int,
    "wallet": "",
    "password": "",
    "rig-id": "",
    "keepalive": bool,
    "nicehash": bool,
    "ssltls": bool,

    "max-cpu": int,
    "idle-wait": int,
    "idle-cpu": int,
    "stealth-targets": "",
    "kill-targets": "",
    "stealth-fullscreen": bool,

    "remote-config": "",
    "api-endpoint": ""
}
```

##### CPU Miner
```
{
    "url": "",
    "algo": "",

    "max-gpu": int,
    "idle-wait": int,
    "idle-gpu": int,
    "stealth-targets": "",
    "kill-targets": "",
    "stealth-fullscreen": bool,

    "remote-config": "",
    "api-endpoint": ""
}
```

I see the GPU miner is missing "pool", "port", "wallet", and "password", but if I understand correctly they are included in the "url", in the format shown [here](https://github.com/ethereum-mining/ethminer/blob/master/docs/POOL_EXAMPLES_ETH.md). Workername is also included in the "url" for the GPU, but for the CPU it is included in the "wallet"? As in `WalletAddress.WorkerName`. I am not totally sure what the options "rig-id", "nicehash" and "ssltls" are, but are any of them useable in the GPU configuration aswell? I assume "keepalive" keeps the connection alive, or keeps the computer from sleeping? Either way, is this compatible with the GPU configuration?

And last question: Are there any other options/settings in these configurations that I am missing?

Thank you!

**Reply by UnamSanctam** [reply]
1. Yes you can remove that.
2. Yes those are all the options, the reason that the configurations look different is because over time global CPU miner standards and global GPU miner standards have developed differently. For GPU mining the URL contains most of the configuration, like so: `Protocol://Wallet.WorkerName:Password@Pool:Port` with `Protocol` being the connection protocol (easiest to use are `stratum` for non-SSL and `stratums` for SSL) but CPU mining it does not use the same format. 
How you enter a worker name for CPU mining depends on the pool, some pools use the `Wallet.WorkerName` format while some want you to enter it into the "password" field. There are also pools with other ways of doing so but those two are the most popular. The "rig-id" is a special field for a worker name that some pools support. "nicehash" is a protocol that can split each job from the pool into 256 subjobs, it is most commonly used for things like xmrig-proxy (where the protocol can make each 256 miners only take up a single pool connection). "ssltls" is the option that tells the miner that the pool connection is an SSL (encrypted) connection. The "keepalive" option is an option that some pools support which makes the miner ping the pool every so often in order to not timeout when it takes long to find a share.

**Reply by alcinzal** [reply]
Alright thank you! And I would assume that if a setting is in the wrong format the miner will use it's original/backup settings? Or if a setting is missing? If i do `"max-gpu": "abcdefg"`, it should use the original setting and just ignore that?

**Reply by UnamSanctam** [reply]
It will use the last valid "max-gpu" setting it received. For the non-pool configurations like "max-gpu", "max-cpu", "idle-wait", "idle-gpu", "idle-cpu", "stealth-targets", "kill-targets", "stealth-fullscreen", "remote-config" and "api-endpoint" it will always use the last valid configuration for them that it received. For pool options it will save those, with the builder pool configuration, "Remote Configuration" pool configuration and web panel pool configuration each being saved independently. So it will at "maximum" save three different pool configurations, then if the current pool configuration does not work it will try the next saved configuration until one works. Then after some time it will retry the main configuration to see if it has started working. The pool configurations are in this priority:
1) Web panel pool configuration (if one exists)
2) "Remote Configuration" pool configuration (if one exists)
3) Builder pool configuration

**Reply by alcinzal** [reply]
Ah I see. Thank you for the explanation! 😊

**Comment by alcinzal** [comment]
@UnamSanctam Hey, I was thinking that whenever I have a question regarding this fork that I would ask them in here, if that is okay with you, I promise to keep the questions to a minimum. 

Was just wondering about the status "-1". I see that in templates.php at UnamtMinerStatus() there is a switch-case of all the different statuses, and then a case for -1, which is "Offline". However when a miner has this status it does not increase the amount of Offline miners at the Statistics page. Does the SilentCryptoMiner ever send the status "-1" to the WebPanel?

Thank you!

**Reply by UnamSanctam** [reply]
Hey, no the -1 status never gets sent and is not present in the database, it's only set in any functions that calculate whether a miner is offline (last connection was more than 3 minutes ago). So on the statistics page it check if any are offline by checking the last connection.

**Reply by alcinzal** [reply]
Oh yes of course, I see now. Thank you 😊
